app/src/lib/git/merge.ts

```diff
@@ -9,15 +13,73 @@ export async function merge(
   await git(['merge', branch], repository.path, 'merge')
 }
 
+/**
+ * Find the base commit between two refs
+ *
+ * @returns the commit id of the merge base, or null if the two refs do not
+ *          have a common base
+ */
 export async function getMergeBase(
   repository: Repository,
   firstRef: string,
   secondRef: string
-): Promise<string> {
+): Promise<string | null> {
   const process = await git(
     ['merge-base', firstRef, secondRef],
     repository.path,
-    'merge-base'
+    'merge-base',
+    {
+      successExitCodes: new Set([0, 1]),
+    }
   )
+
+  if (process.exitCode === 1) {
+    return null
+  }
+
   return process.stdout.trim()
 }
+
+/**
+ * Generate the merge result from two branches in a repository
+ *
+ * @param repository The repository containing the branches to merge
+ * @param ours The current branch
+ * @param theirs Another branch to merge into the current branch
+ */
+export async function mergeTree(
+  repository: Repository,
+  ours: Branch,
+  theirs: Branch
+): Promise<MergeResult | null> {
+  console.time('getMergeBase')
+  const mergeBase = await getMergeBase(repository, ours.tip.sha, theirs.tip.sha)
+  console.timeEnd('getMergeBase')
+
+  if (mergeBase === null) {
+    // TODO: return a better result to indicate something is wrong
+    return null
```

Comment:
Could we throw an error here or will you change this as a part of some other work you have planned?

Reply:
This is handled in #5495 because of a corner case I wrote more about here: https://github.com/desktop/desktop/issues/4588#issuecomment-417382948

```
  if (mergeBase === null) {
    return { kind: MergeResultKind.Invalid }
  }
```